python/ray/external_storage.py

```diff
@@ -35,31 +125,65 @@ def _put_object_to_store(self, metadata, data_size, file_like, object_ref):
         worker.core_worker.put_file_like_object(metadata, data_size, file_like,
                                                 object_ref)
 
+    def _fusion_objects(self, f, object_refs, object_id_stored,
+                        url_protocol) -> List[str]:
+        """Fusion all objects into a given file handle/return internal urls.
+
+        Args:
+            f: File handle to fusion all given object refs.
+            object_refs: Object references to fusion to a single file.
+            object_id_stored: The ID that will be used to store the file f.
+            url_protocol: instance of ExternalStorageURLProtocol.
+
+        Return:
+            List of internal urls of fusioned objects.
+        """
+        keys = []
+        offsets = 0
+        ray_object_pairs = self._get_objects_from_store(object_refs)
+        for ref, (buf, metadata) in zip(object_refs, ray_object_pairs):
+            metadata_len = len(metadata)
+            buf_len = len(buf)
+            object_size_in_bytes = metadata_len + buf_len + 16
+            internal_url = url_protocol.create_internal_url(
+                job_id=ref.job_id().hex(),
+                object_id_stored=object_id_stored,
+                object_id=ref.hex(),
+                object_size_in_bytes=object_size_in_bytes,
+                offsets_in_bytes=offsets)
+            f.write(metadata_len.to_bytes(8, byteorder="little"))
+            f.write(buf_len.to_bytes(8, byteorder="little"))
+            f.write(metadata)
+            f.write(memoryview(buf))
+            offsets += object_size_in_bytes
+            keys.append(internal_url.encode())
+        return keys
+
     @abc.abstractmethod
-    def spill_objects(self, object_refs):
+    def spill_objects(self, object_refs) -> List[str]:
         """Spill objects to the external storage. Objects are specified
         by their object refs.
 
         Args:
             object_refs: The list of the refs of the objects to be spilled.
         Returns:
-            A list of keys corresponding to the input object refs.
+            A list of internal URLs following ExternalStorageURLProtocol.
         """
 
     @abc.abstractmethod
-    def restore_spilled_objects(self, keys: List[bytes]):
+    def restore_spilled_objects(self, keys: List[str]):
         """Spill objects to the external storage. Objects are specified
         by their object refs.
 
         Args:
```

Comment:
In the future, we should support read-ahead to return extra objects we happened to read in the same chunk. Not needed for now though.